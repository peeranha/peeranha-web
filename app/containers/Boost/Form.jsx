import React, { memo } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { reduxForm } from 'redux-form/immutable';
import { injectIntl, FormattedMessage } from 'react-intl';

import { scrollToErrorField } from 'utils/animation';
import { singleCommunityStyles } from 'utils/communityManagement';

import { DARK_SECONDARY, BORDER_RADIUS_M } from 'style-constants';

import TipsBase from 'components/Base/TipsBase';
import { BaseSpecialOne } from 'components/Base/BaseTransparent';
import FormBox from 'components/Form';
import Button from 'components/Button/Contained/InfoLarge';
import TransparentButton from 'components/Button/Contained/Transparent';
import Label from 'components/FormFields/Label';
import BlockedInfoArea from 'components/BlockedInfoArea';
import Tips from './Tips';
import PredictionForm from './PredictionForm';
import CurrentStakeForm, { MinStake } from './CurrentStakeForm';

import { FORM_TYPE, CURRENT_STAKE_FORM } from './constants';

import messages from './messages';

const { projectBorderRadius } = singleCommunityStyles();

export const InputWrapper = styled.div`
  position: relative;
  margin-bottom: 40px;

  ${Label} {
    margin-bottom: 10px;
    font-size: 16px;
    font-weight: bold;
  }

  input {
    height: 45px;
    font-size: 20px;
    font-weight: bold;

    ${({ isPrediction }) => (isPrediction ? `:disabled { opacity: 1; }` : '')};
  }

  .err + ${MinStake} {
    display: none;
  }

  .err + span + span + div {
    display: none;
  }
`;

export const InputProgressBar = styled.div`
  position: absolute;
  bottom: 0;
  left: 0;

  width: ${({ width }) =>
    projectBorderRadius
      ? `calc(${width}% - ${BORDER_RADIUS_M})` || '0%'
      : `${width || 0}%`};

  margin-left: ${() =>
    projectBorderRadius
      ? `${Number(BORDER_RADIUS_M.split('px')[0]) * 0.75}px`
      : ''};

  max-width: 100%;
  height: 4px;
  background-color: ${DARK_SECONDARY};
`;

const Title = styled.p`
  margin-top: 15px;
  margin-bottom: 25px;
  margin-left: 25px;
  font-size: 30px;
  font-weight: 600;
  line-height: 1.2;

  @media (max-width: 768px) {
    margin-top: 20px;
    margin-bottom: 15px;
    margin-left: 15px;
    font-size: 22px;
  }
`;

const Form = ({
  handleSubmit,
  reset,
  changeStake,
  changeStakeLoading,
  changeCurrentStake,
  maxStake,
  initialUserStake,
  currentStake,
  onChangeCurrentStake,
  locale,
  formValues,
  nextWeekMaxStake,
}) => {
  const resetAction =
    initialUserStake === currentStake
      ? reset
      : () => onChangeCurrentStake(initialUserStake || 0);

  return (
    <div className="mb-5">
      {formValues[CURRENT_STAKE_FORM] !== undefined && (
        <>
          <Title>
            <FormattedMessage {...messages.formTitle} />
          </Title>
          <TipsBase>
            <BaseSpecialOne className="position-relative">
              {!maxStake && (
                <BlockedInfoArea>
                  <FormattedMessage {...messages.notTokensToStake} />
                </BlockedInfoArea>
              )}

              <FormBox onSubmit={handleSubmit(changeStake)}>
                <PredictionForm
                  locale={locale}
                  formValues={formValues}
                  maxStake={nextWeekMaxStake}
                />

                <CurrentStakeForm
                  maxValue={maxStake}
                  onClickStakeTag={v => changeCurrentStake(v)}
                  disabled={changeStakeLoading}
                  formValues={formValues}
                  formSubmitAction={handleSubmit(changeStake)}
                  initialUserStake={initialUserStake}
                />

                <div className="mt-5">
                  <Button type="submit" className="mr-4">
                    <FormattedMessage {...messages.formSubmit} />
                  </Button>
                  <TransparentButton type="reset" onClick={resetAction}>
                    <FormattedMessage {...messages.formCancel} />
                  </TransparentButton>
                </div>
              </FormBox>
            </BaseSpecialOne>

            <Tips />
          </TipsBase>
        </>
      )}
    </div>
  );
};

Form.propTypes = {
  handleSubmit: PropTypes.func,
  reset: PropTypes.func,
  formValues: PropTypes.object,
  changeStake: PropTypes.func,
  changeStakeLoading: PropTypes.bool,
  changeCurrentStake: PropTypes.func,
  maxStake: PropTypes.number,
  initialUserStake: PropTypes.number,
  onChangeCurrentStake: PropTypes.func,
  locale: PropTypes.string,
  currentStake: PropTypes.number,
  nextWeekMaxStake: PropTypes.number,
};

const FormClone = reduxForm({
  form: FORM_TYPE,
  onSubmitFail: errors => scrollToErrorField(errors),
})(Form);

export default memo(
  injectIntl(
    connect((state, { currentStake }) => {
      const form = state.toJS().form[FORM_TYPE] || { values: {} };

      const currentStakeValue =
        typeof currentStake === 'number' && currentStake > 0 ? currentStake : 0;

      return {
        formValues: form.values,
        initialValues: {
          [CURRENT_STAKE_FORM]: currentStakeValue.toString(),
        },
        enableReinitialize: true,
      };
    })(FormClone),
  ),
);
