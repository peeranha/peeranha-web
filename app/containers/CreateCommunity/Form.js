import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Field, reduxForm, FormSection } from 'redux-form/immutable';
import { FormattedMessage } from 'react-intl';

import icoTag from 'images/icoTag.svg';
import closeIcon from 'svg/close';

import Wrapper from 'components/FormFields/Wrapper';
import TextareaField from 'components/FormFields/TextareaField';
import TextInputField from 'components/FormFields/TextInputField';
import AvatarField from 'components/FormFields/AvatarField';
import LargeButton from 'components/Button/Contained/InfoLarge';
import TransparentButton from 'components/Button/Contained/Transparent';
import Icon from 'components/Icon';

import FormStyled from 'containers/EditProfilePage/FormStyled';
import AvatarStyled from 'containers/EditProfilePage/AvatarStyled';

import {
  AVATAR_FIELD_WIDTH,
  AVATAR_FIELD_MARGIN,
} from 'containers/EditProfilePage/ProfileEditForm';

import {
  imageValidation,
  strLength3x20,
  required,
  strLength20x1000,
} from 'components/FormFields/validate';

import messages from './messages';

import {
  COMM_AVATAR_FIELD,
  COMM_NAME_FIELD,
  COMM_SHORT_DESCRIPTION_FIELD,
  COMM_MAIN_DESCRIPTION_FIELD,
  TAG_NAME_FIELD,
  TAG_DESCRIPTION_FIELD,
  TAG_SECTION,
} from './constants';

const DEFAULT_TAGS_NUMBER = 3;
const DEFAULT_TAGS_ARRAY = [];

/* eslint no-plusplus: 0 */
for (let i = 0; i < DEFAULT_TAGS_NUMBER; i++) {
  DEFAULT_TAGS_ARRAY.push(i);
}

/* eslint-disable-next-line */
export const CreateCommunityForm = /* istanbul ignore next */ ({
  handleSubmit,
  createCommunity,
  createCommunityLoading,
  translations,
  change,
  editingImgState,
  uploadImage,
  cachedProfileImg,
  getCroppedAvatar,
  clearImageChanges,
}) => {
  const [tags, changeTags] = useState(DEFAULT_TAGS_ARRAY);

  const removeTag = e => {
    if (tags.length === 1) return;

    const { key } = e.currentTarget.dataset;
    const index = tags.findIndex(x => x === +key);

    // clear redux store
    change(`${TAG_NAME_FIELD}_${key}`, null);
    change(`${TAG_DESCRIPTION_FIELD}_${key}`, null);

    // clear array
    const tagsCopy = [...tags];
    tagsCopy.splice(index, 1);

    changeTags(tagsCopy);
  };

  const addTag = () => {
    changeTags([...tags, tags[tags.length - 1] + 1]);
  };

  return (
    <FormStyled
      size={AVATAR_FIELD_WIDTH + AVATAR_FIELD_MARGIN}
      onSubmit={handleSubmit(createCommunity)}
    >
      <div className="position-static">
        <AvatarStyled>
          <Field
            editingImgState={editingImgState}
            uploadImage={uploadImage}
            cachedProfileImg={cachedProfileImg}
            ipfsAvatar={null}
            getCroppedAvatar={getCroppedAvatar}
            clearImageChanges={clearImageChanges}
            disabled={createCommunityLoading}
            name={COMM_AVATAR_FIELD}
            label={translations[messages.avatar.id]}
            component={AvatarField}
            validate={imageValidation}
            warn={imageValidation}
            size={AVATAR_FIELD_WIDTH}
          />
        </AvatarStyled>

        <Field
          disabled={createCommunityLoading}
          name={COMM_NAME_FIELD}
          component={TextInputField}
          label={translations[messages.communityTitle.id]}
          validate={[strLength3x20, required]}
          warn={[strLength3x20, required]}
          tip={translations[messages.communityTitleTip.id]}
        />

        <Field
          disabled={createCommunityLoading}
          name={COMM_SHORT_DESCRIPTION_FIELD}
          component={TextInputField}
          label={translations[messages.shortDescription.id]}
          validate={[strLength3x20, required]}
          warn={[strLength3x20, required]}
          tip={translations[messages.shortDescriptionTip.id]}
        />

        <Field
          disabled={createCommunityLoading}
          name={COMM_MAIN_DESCRIPTION_FIELD}
          component={TextareaField}
          label={translations[messages.whyWeNeedIt.id]}
          validate={[strLength20x1000, required]}
          warn={[strLength20x1000, required]}
          tip={translations[messages.whyWeNeedItTip.id]}
        />

        <div>
          <Wrapper label={translations[messages.tags.id]} tip>
            <FormattedMessage {...messages.tagsAreNeeded} />
          </Wrapper>

          <FormSection name="tags" className="mt-3">
            {tags.map(x => (
              <FormSection name={`${TAG_SECTION}_${x}`}>
                {x !== tags[0] && (
                  <button
                    type="button"
                    style={{ transform: 'scale(0.75)' }}
                    data-key={x}
                    onClick={removeTag}
                  >
                    <Icon icon={closeIcon} />
                  </button>
                )}

                <Field
                  disabled={createCommunityLoading}
                  name={`${TAG_NAME_FIELD}_${x}`}
                  component={TextInputField}
                  placeholder={translations[messages.tagTitle.id]}
                  validate={[strLength3x20, required]}
                  warn={[strLength3x20, required]}
                  tip={translations[messages.tagTitleTip.id]}
                />

                <Field
                  disabled={createCommunityLoading}
                  name={`${TAG_DESCRIPTION_FIELD}_${x}`}
                  component={TextareaField}
                  placeholder={translations[messages.tagDescription.id]}
                  validate={[strLength20x1000, required]}
                  warn={[strLength20x1000, required]}
                  tip={translations[messages.tagDescriptionTip.id]}
                />
              </FormSection>
            ))}
          </FormSection>
        </div>

        <div>
          <TransparentButton
            className="d-flex align-items-center"
            type="button"
            onClick={addTag}
          >
            <img className="mr-2" src={icoTag} alt="icoTag" />
            <FormattedMessage {...messages.oneMoreTag} />
          </TransparentButton>
        </div>

        <div>
          <LargeButton
            className="my-3"
            disabled={createCommunityLoading}
            typeAttr="submit"
          >
            {translations[messages.createCommunity.id]}
          </LargeButton>
        </div>
      </div>
    </FormStyled>
  );
};

CreateCommunityForm.propTypes = {
  handleSubmit: PropTypes.func,
  createCommunity: PropTypes.func,
  createCommunityLoading: PropTypes.bool,
  translations: PropTypes.object,
  invalid: PropTypes.bool,
  submitting: PropTypes.bool,
  editingImgState: PropTypes.bool,
  cachedProfileImg: PropTypes.string,
  getCroppedAvatar: PropTypes.func,
  uploadImage: PropTypes.func,
  change: PropTypes.func,
  clearImageChanges: PropTypes.func,
};

export default reduxForm({
  form: 'CreateCommunityForm',
})(CreateCommunityForm);
