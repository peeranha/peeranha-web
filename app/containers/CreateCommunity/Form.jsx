import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Field, reduxForm, FormSection } from 'redux-form/immutable';
import { FormattedMessage } from 'react-intl';

import { appLocales } from 'i18n';

import icoTag from 'images/icoTag.svg?inline';
import closeIcon from 'images/close.svg?inline';

import Wrapper from 'components/FormFields/Wrapper';
import TextareaField from 'components/FormFields/TextareaField';
import TextInputField from 'components/FormFields/TextInputField';
import AvatarField from 'components/FormFields/AvatarField';

import SelectField, {
  getSelectOptions,
} from 'components/FormFields/SelectField';

import LargeButton from 'components/Button/Contained/InfoLarge';
import TransparentButton from 'components/Button/Contained/Transparent';

import FormStyled from 'containers/EditProfilePage/FormStyled';
import AvatarStyled from 'containers/EditProfilePage/AvatarStyled';

import {
  AVATAR_FIELD_WIDTH,
  AVATAR_FIELD_MARGIN,
} from 'containers/EditProfilePage/ProfileEditForm';

import {
  strLength3x20,
  required,
  strLength20x1000,
  strLength15x100,
  imageValidation,
} from 'components/FormFields/validate';

import messages from './messages';

import {
  FORM_NAME,
  COMM_AVATAR_FIELD,
  COMM_NAME_FIELD,
  COMM_SHORT_DESCRIPTION_FIELD,
  COMM_MAIN_DESCRIPTION_FIELD,
  TAG_NAME_FIELD,
  TAG_DESCRIPTION_FIELD,
  TAG_SECTION,
  LANGUAGE_FIELD,
  CREATE_COMMUNITY_BUTTON,
} from './constants';

const DEFAULT_TAGS_NUMBER = 5;
const DEFAULT_TAGS_ARRAY = [];

/* eslint no-plusplus: 0 */
for (let i = 0; i < DEFAULT_TAGS_NUMBER; i++) {
  DEFAULT_TAGS_ARRAY.push(i);
}

/* eslint react/jsx-no-bind: 0 */
/* eslint-disable-next-line */
const CreateCommunityForm = ({
  handleSubmit,
  createCommunity,
  createCommunityLoading,
  translations,
  change,
}) => {
  const [tags, changeTags] = useState(DEFAULT_TAGS_ARRAY);

  const removeTag = e => {
    if (tags.length === 5) return;

    const { key } = e.currentTarget.dataset;
    const index = tags.findIndex(x => x === +key);

    // clear tag in redux-form
    change(`tags.${TAG_SECTION}_${key}`, null);

    // clear array
    const tagsCopy = [...tags];
    tagsCopy.splice(index, 1);

    changeTags(tagsCopy);
  };

  const addTag = () => {
    changeTags([...tags, tags[tags.length - 1] + 1]);
  };

  return (
    <FormStyled
      size={AVATAR_FIELD_WIDTH + AVATAR_FIELD_MARGIN}
      onSubmit={handleSubmit(createCommunity)}
    >
      <div className="position-static">
        <AvatarStyled>
          <Field
            name={COMM_AVATAR_FIELD}
            component={AvatarField}
            size={AVATAR_FIELD_WIDTH}
            validate={[imageValidation, required]}
            warn={[imageValidation, required]}
            disabled={createCommunityLoading}
          />
        </AvatarStyled>

        <Field
          disabled={createCommunityLoading}
          name={COMM_NAME_FIELD}
          component={TextInputField}
          label={translations[messages.communityTitle.id]}
          validate={[strLength3x20, required]}
          warn={[strLength3x20, required]}
          tip={translations[messages.communityTitleTip.id]}
        />

        <Field
          name={LANGUAGE_FIELD}
          placeholder=""
          options={getSelectOptions(appLocales)}
          label={translations[messages.communityLanguage.id]}
          tip={translations[messages.communityLanguageTip.id]}
          disabled={createCommunityLoading}
          component={SelectField}
          validate={[required]}
          warn={[required]}
        />

        <Field
          disabled={createCommunityLoading}
          name={COMM_SHORT_DESCRIPTION_FIELD}
          component={TextInputField}
          label={translations[messages.shortDescription.id]}
          validate={[strLength15x100, required]}
          warn={[strLength15x100, required]}
          tip={translations[messages.shortDescriptionTip.id]}
        />

        <Field
          disabled={createCommunityLoading}
          name={COMM_MAIN_DESCRIPTION_FIELD}
          component={TextareaField}
          label={translations[messages.whyWeNeedIt.id]}
          validate={[strLength20x1000, required]}
          warn={[strLength20x1000, required]}
          tip={translations[messages.whyWeNeedItTip.id]}
        />

        <div>
          <Wrapper label={translations[messages.tags.id]} tip>
            <FormattedMessage {...messages.tagsAreNeeded} />
          </Wrapper>

          <FormSection name="tags" className="mt-3">
            {tags.map((x, index) => (
              <FormSection key={x} name={`${TAG_SECTION}_${x}`}>
                {index >= DEFAULT_TAGS_NUMBER && (
                  <button
                    type="button"
                    style={{ transform: 'scale(0.75)' }}
                    data-key={x}
                    onClick={removeTag}
                  >
                    <img src={closeIcon} alt="icon" />
                  </button>
                )}

                <Field
                  disabled={createCommunityLoading}
                  name={TAG_NAME_FIELD}
                  component={TextInputField}
                  placeholder={translations[messages.tagTitle.id]}
                  validate={[strLength3x20, required]}
                  warn={[strLength3x20, required]}
                  tip={translations[messages.tagTitleTip.id]}
                />

                <Field
                  disabled={createCommunityLoading}
                  name={TAG_DESCRIPTION_FIELD}
                  component={TextareaField}
                  placeholder={translations[messages.tagDescription.id]}
                  validate={[strLength20x1000, required]}
                  warn={[strLength20x1000, required]}
                  tip={translations[messages.tagDescriptionTip.id]}
                />
              </FormSection>
            ))}
          </FormSection>
        </div>

        <div>
          <TransparentButton
            className="d-flex align-items-center"
            type="button"
            onClick={addTag}
          >
            <img className="mr-2" src={icoTag} alt="icoTag" />
            <FormattedMessage {...messages.oneMoreTag} />
          </TransparentButton>
        </div>

        <div>
          <LargeButton
            className="my-3"
            disabled={createCommunityLoading}
            type="submit"
            id={CREATE_COMMUNITY_BUTTON}
          >
            {translations[messages.createCommunity.id]}
          </LargeButton>
        </div>
      </div>
    </FormStyled>
  );
};

CreateCommunityForm.propTypes = {
  handleSubmit: PropTypes.func,
  createCommunity: PropTypes.func,
  createCommunityLoading: PropTypes.bool,
  translations: PropTypes.object,
  invalid: PropTypes.bool,
  submitting: PropTypes.bool,
  change: PropTypes.func,
};

const validateTagsTitles = st => {
  const state = st.toJS();
  const errors = {
    tags: {},
  };

  if (!state.tags) return errors;

  const tags = Object.keys(state.tags).filter(x => state.tags[x]);

  const notValidTitles = tags.filter(
    x =>
      tags
        .map(z => state.tags[z][TAG_NAME_FIELD])
        .filter(y => y === state.tags[x][TAG_NAME_FIELD]).length > 1,
  );

  notValidTitles.forEach(x => {
    errors.tags[x] = {
      [TAG_NAME_FIELD]: { id: messages.onlyTagsWithUniqueTitles.id },
    };
  });

  return errors;
};

export { CreateCommunityForm, validateTagsTitles };

export default reduxForm({
  form: FORM_NAME,
  validate: (state, props) => validateTagsTitles(state, props),
  warn: (state, props) => validateTagsTitles(state, props),
})(CreateCommunityForm);
