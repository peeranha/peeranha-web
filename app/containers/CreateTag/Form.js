import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Field, reduxForm } from 'redux-form/immutable';

import { getFollowedCommunities } from 'utils/communityManagement';

import TextareaField from 'components/FormFields/TextareaField';
import TextInputField from 'components/FormFields/TextInputField';
import CommunityField from 'components/FormFields/CommunityField';
import Button from 'components/Button/Contained/InfoLarge';

import {
  required,
  strLength2x15,
  strLength20x1000,
  requiredForObjectField,
} from 'components/FormFields/validate';

import {
  NAME_FIELD,
  DESCRIPTION_FIELD,
  FORM_COMMUNITY,
  FORM_NAME,
} from './constants';

import messages from './messages';

export const Form = /* istanbul ignore next */ ({
  createTagLoading,
  createTag,
  handleSubmit,
  translations,
  communities,
  formValues,
  change,
}) => {
  change(FORM_COMMUNITY, formValues[FORM_COMMUNITY]);

  return (
    <form onSubmit={handleSubmit(createTag)}>
      <div>
        <Field
          name={FORM_COMMUNITY}
          component={CommunityField}
          disabled={createTagLoading}
          label={translations[messages.community.id]}
          tip={translations[messages.communityTip.id]}
          options={communities}
          validate={[requiredForObjectField]}
          warn={[requiredForObjectField]}
          splitInHalf
        />
        <Field
          disabled={createTagLoading}
          name={NAME_FIELD}
          component={TextInputField}
          label={translations[messages.name.id]}
          tip={translations[messages.nameTip.id]}
          validate={[strLength2x15, required]}
          warn={[strLength2x15, required]}
          splitInHalf
        />
        <Field
          disabled={createTagLoading}
          name={DESCRIPTION_FIELD}
          component={TextareaField}
          label={translations[messages.descriptionField.id]}
          tip={translations[messages.descriptionFieldTip.id]}
          validate={[strLength20x1000, required]}
          warn={[strLength20x1000, required]}
          splitInHalf
        />
      </div>

      <div>
        <Button className="my-3" disabled={createTagLoading}>
          {translations[messages.createTag.id]}
        </Button>
      </div>
    </form>
  );
};

Form.propTypes = {
  createTagLoading: PropTypes.bool,
  createTag: PropTypes.func,
  handleSubmit: PropTypes.func,
  translations: PropTypes.object,
  communities: PropTypes.array,
  formValues: PropTypes.object,
  change: PropTypes.func,
};

let FormClone = reduxForm({
  form: FORM_NAME,
})(Form);

FormClone = connect(
  /* istanbul ignore next */ (state, props) => {
    const { communities, communityId } = props;

    const form = state.toJS().form[FORM_NAME];

    return {
      formValues: form ? form.values : {},
      initialValues: {
        [FORM_COMMUNITY]: getFollowedCommunities(communities, [communityId])[0],
      },
    };
  },
)(FormClone);

export default React.memo(FormClone);
