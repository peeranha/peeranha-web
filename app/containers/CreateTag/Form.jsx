import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Field, reduxForm } from 'redux-form/immutable';

import {
  getFollowedCommunities,
  isSingleCommunityWebsite,
} from 'utils/communityManagement';
import { scrollToErrorField } from 'utils/animation';

import TextareaField from 'components/FormFields/TextareaField';
import TextInputField from 'components/FormFields/TextInputField';
import CommunityField from 'components/FormFields/CommunityField';
import Button from 'components/Button/Contained/InfoLarge';
import FormBox from 'components/Form';

import {
  required,
  strLength2x15,
  strLength20x1000,
  requiredForObjectField,
  valueHasNotBeInList,
} from 'components/FormFields/validate';

import {
  NAME_FIELD,
  DESCRIPTION_FIELD,
  FORM_COMMUNITY,
  FORM_NAME,
} from './constants';

import messages from './messages';

export const Form = ({
  createTagLoading,
  createTag,
  handleSubmit,
  translations,
  communities,
}) => (
  <FormBox onSubmit={handleSubmit(createTag)}>
    <Field
      className={isSingleCommunityWebsite() ? 'd-none' : ''}
      name={FORM_COMMUNITY}
      component={CommunityField}
      disabled={createTagLoading}
      label={translations[messages.community.id]}
      tip={translations[messages.communityTip.id]}
      options={communities}
      validate={[requiredForObjectField]}
      warn={[requiredForObjectField]}
      splitInHalf
    />

    <Field
      disabled={createTagLoading}
      name={NAME_FIELD}
      component={TextInputField}
      label={translations[messages.name.id]}
      tip={translations[messages.nameTip.id]}
      validate={[strLength2x15, required, valueHasNotBeInList]}
      warn={[strLength2x15, required, valueHasNotBeInList]}
      splitInHalf
    />

    <Field
      disabled={createTagLoading}
      name={DESCRIPTION_FIELD}
      component={TextareaField}
      label={translations[messages.descriptionField.id]}
      tip={translations[messages.descriptionFieldTip.id]}
      validate={[strLength20x1000, required]}
      warn={[strLength20x1000, required]}
      splitInHalf
    />

    <Button type="submit" disabled={createTagLoading}>
      {translations[messages.createTag.id]}
    </Button>
  </FormBox>
);

Form.propTypes = {
  createTagLoading: PropTypes.bool,
  createTag: PropTypes.func,
  handleSubmit: PropTypes.func,
  translations: PropTypes.object,
  communities: PropTypes.array,
};

let FormClone = reduxForm({
  form: FORM_NAME,
  onSubmitFail: errors => scrollToErrorField(errors),
})(Form);

FormClone = connect((state, props) => {
  const { communities, communityId } = props;

  const form = state.toJS().form[FORM_NAME];

  return {
    valueHasNotBeInListValidate: (
      (form &&
        form.values &&
        form.values[FORM_COMMUNITY] &&
        form.values[FORM_COMMUNITY].tags) ||
      []
    ).map(x => x.name),
    initialValues: {
      [FORM_COMMUNITY]: getFollowedCommunities(communities, [communityId])[0],
    },
  };
})(FormClone);

export default React.memo(FormClone);
