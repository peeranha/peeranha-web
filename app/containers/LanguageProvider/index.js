/*
 *
 * LanguageProvider
 *
 * this component connects the redux state language locale to the
 * IntlProvider component and i18n messages (loaded from `app/translations`)
 */

import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createSelector } from 'reselect';
import { IntlProvider } from 'react-intl';
import { bindActionCreators } from 'redux';
import { translationMessages } from 'i18n';

import { makeSelectLocale } from './selectors';
import { changeLocale } from './actions';

/* eslint react/no-did-mount-set-state: 0 prefer-destructuring: 0  */
export class LanguageProvider extends React.PureComponent {
  componentWillMount() {
    const languages = Object.keys(translationMessages);
    let locale = localStorage.getItem('locale');

    // if (!locale) - find the first suitable language in window.navigator.languages
    if (!locale) {
      locale = window.navigator.languages.filter(x => languages.includes(x))[0];
    }

    if (locale) {
      this.props.changeLocaleDispatch(locale);
    }
  }

  // eslint-disable-line react/prefer-stateless-function
  render() {
    return (
      <IntlProvider
        locale={this.props.locale}
        key={this.props.locale}
        messages={this.props.messages[this.props.locale]}
      >
        {React.Children.only(this.props.children)}
      </IntlProvider>
    );
  }
}

LanguageProvider.propTypes = {
  locale: PropTypes.string,
  messages: PropTypes.object,
  children: PropTypes.element,
  changeLocaleDispatch: PropTypes.func,
};

const mapStateToProps = createSelector(makeSelectLocale(), locale => ({
  locale,
}));

function mapDispatchToProps(dispatch) /* istanbul ignore next */ {
  return {
    changeLocaleDispatch: bindActionCreators(changeLocale, dispatch),
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(LanguageProvider);
