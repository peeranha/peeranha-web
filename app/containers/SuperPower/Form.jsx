import React, { memo, useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { Field, reduxForm } from 'redux-form/immutable';
import { injectIntl, FormattedMessage } from 'react-intl';

import { scrollToErrorField } from 'utils/animation';

import messages from './messages';

import {
  FORM_TYPE,
  BET_TYPE_FORM,
  CURRENT_BET_FORM,
  SUPERPOWER_PREDICTION_FORM,
} from './constants';
import {
  BG_LIGHT,
  BG_WARNING_LIGHT_TRANSPARENT,
  SECONDARY_SPECIAL,
} from 'style-constants';

import TipsBase from 'components/Base/TipsBase';
import { BaseSpecialOne } from 'components/Base/BaseTransparent';
import Tips from './Tips';
import PredictionForm from './PredictionForm';
import CurrentBetForm from './CurrentBetForm';
import FormBox from 'components/Form';
import Button from 'components/Button/Contained/InfoLarge';
import Checkbox from 'components/Input/Checkbox';
import Label from 'components/FormFields/Label';

export const InputWrapper = styled.div`
  position: relative;
  margin-bottom: 40px;

  ${Label} {
    margin-bottom: 10px;
    font-size: 16px;
    font-weight: bold;
  }

  input {
    height: 45px;
    font-size: 20px;
    font-weight: bold;
    ${({ extraPosition }) => extraPosition ? `
      background: linear-gradient(90deg, ${BG_LIGHT} ${extraPosition}%, ${BG_WARNING_LIGHT_TRANSPARENT} ${extraPosition}%);

      :disabled {
        opacity: 1;
      }
    ` : ''}
  }
`;

export const InputProgressBar = styled.div`
  position: absolute;
  bottom: 0;
  left: 0;
  width: ${({ width }) => width || 0}%;
  height: 4px;
  background-color: ${SECONDARY_SPECIAL};
`;

const Title = styled.p`
  margin-top: 15px;
  margin-bottom: 25px;
  margin-left: 25px;
  font-size: 30px;
  font-weight: 600;
  line-height: 1.2;

  @media (max-width: 768px) {
    margin-top: 20px;
    margin-bottom: 15px;
    margin-left: 15px;
    font-size: 22px;
  }
`;

const Form = ({
  handleSubmit,
  formValues,
  currentBetValue,
  changeBet,
  changeBetLoading,
  changeCurrentBet,
  superPowerPrediction,
  maxBet,
  onChangeCurrentBet,
  locale,
}) => {

  return (
    <>
      <Title><FormattedMessage {...messages.formTitle} /></Title>
      <TipsBase>
        <BaseSpecialOne>
          <FormBox onSubmit={handleSubmit(changeBet)}>
            <PredictionForm value={superPowerPrediction} locale={locale} />

            <CurrentBetForm
              maxValue={maxBet}
              value={+currentBetValue}
              onClickBetTag={v => changeCurrentBet(v)}
              disabled={changeBetLoading}
              onChange={onChangeCurrentBet}
            />

            <Field
              name={BET_TYPE_FORM}
              component={Checkbox}
              label={<FormattedMessage {...messages.formBetType} />}
              disabled={changeBetLoading}
            />
            <Button type="submit">
              <FormattedMessage {...messages.formSubmit} />
            </Button>
          </FormBox>
        </BaseSpecialOne>

        <Tips />
      </TipsBase>
    </>
  );
};

Form.propTypes = {
  handleSubmit: PropTypes.func,
  formValues: PropTypes.object,
  currentBetValue: PropTypes.number,
  changeBet: PropTypes.func,
  changeBetLoading: PropTypes.bool,
  changeCurrentBet: PropTypes.func,
  superPowerPrediction: PropTypes.number,
  maxBet: PropTypes.number,
  onChangeCurrentBet: PropTypes.func,
  locale: PropTypes.string,
};

const FormClone = reduxForm({
  form: FORM_TYPE,
  onSubmitFail: errors => scrollToErrorField(errors),
})(Form);

export default memo(
  injectIntl(
    connect(
      (
        state,
        { currentBet, superPowerPrediction }
      ) => {
        const form = state.toJS().form[FORM_TYPE] || { values: {} };

        return {
          formValues: form,
          currentBetValue: +form.values[CURRENT_BET_FORM],
          initialValues: {
            [BET_TYPE_FORM]: form.values[BET_TYPE_FORM] || false,
            [SUPERPOWER_PREDICTION_FORM]: `x${superPowerPrediction}`,
            [CURRENT_BET_FORM]: currentBet.toString(),
          },
          enableReinitialize: true,
        };
      },
      dispatch => ({}),
    )(FormClone),
  ),
);
