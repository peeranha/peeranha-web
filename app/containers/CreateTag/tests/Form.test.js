import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import { memoryHistory } from 'react-router-dom';
import configureStore from 'configureStore';
import { ConnectedRouter } from 'react-router-redux';
import createdHistory from 'createdHistory';

import LanguageProvider from 'containers/LanguageProvider';

import { Form } from '../Form';

describe('<Form />', () => {
  const props = {
    handleSubmit: () => {},
    submitting: true,
    invalid: true,
    createTag: jest.fn(),
    createTagLoading: false,
    translations: {},
  };

  it('snapshot test', () => {
    const store = configureStore({}, memoryHistory);
    const renderedComponent = shallow(
      <Provider store={store}>
        <LanguageProvider locale="en" key="en" messages={{}}>
          <ConnectedRouter history={createdHistory}>
            <Form {...props} />
          </ConnectedRouter>
        </LanguageProvider>
      </Provider>,
    );
    expect(renderedComponent).toMatchSnapshot();
  });
});
