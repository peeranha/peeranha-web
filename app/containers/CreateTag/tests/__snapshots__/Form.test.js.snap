// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> snapshot test 1`] = `
<Connect(LanguageProvider)
  key="en"
  locale="en"
  messages={Object {}}
>
  <ConnectedRouter
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Form
      createTag={[MockFunction]}
      createTagLoading={false}
      handleSubmit={[Function]}
      invalid={true}
      submitting={true}
      translations={Object {}}
    />
  </ConnectedRouter>
</Connect(LanguageProvider)>
`;
